types/src/system/auction/detail.rs

```diff
@@ -205,10 +205,39 @@ pub(crate) fn process_unbond_requests<P: Auction + ?Sized>(provider: &mut P) ->
             // current era id + unbonding delay is equal or greater than the `era_of_creation` that
             // was calculated on `unbond` attempt.
             if current_era_id >= unbonding_purse.era_of_creation() + unbonding_delay {
-                // Move funds from bid purse to unbonding purse
-                provider
-                    .unbond(unbonding_purse)
-                    .map_err(|_| Error::TransferToUnbondingPurse)?;
+                match unbonding_purse.new_validator() {
+                    Some(new_validator) => {
+                        match provider.read_bid(&new_validator.to_account_hash()) {
+                            Ok(Some(bid)) => {
+                                if !bid.staked_amount().is_zero() {
+                                    provider
+                                        .handle_delegation(
+                                            unbonding_purse.unbonder_public_key().clone(),
+                                            new_validator.clone(),
+                                            *unbonding_purse.bonding_purse(),
+                                            *unbonding_purse.amount(),
+                                        )
+                                        .map(|_| ())?
+                                } else {
+                                    // Move funds from bid purse to unbonding purse
+                                    provider
+                                        .unbond(unbonding_purse)
+                                        .map_err(|_| Error::TransferToUnbondingPurse)?
+                                }
+                            }
+                            // Move funds from bid purse to unbonding purse
+                            Ok(_) | Err(_) => provider
+                                .unbond(unbonding_purse)
+                                .map_err(|_| Error::TransferToUnbondingPurse)?,
+                        }
+                    }
+                    None => {
+                        // Move funds from bid purse to unbonding purse
+                        provider
+                            .unbond(unbonding_purse)
+                            .map_err(|_| Error::TransferToUnbondingPurse)?
```

Comment:
It'd be great to reorganize error handling here as `unbond()` is called three times at different indentation levels.

Reply:
Yeah, I can't seem to find a structure that can accomodate that